gk/main.c

```diff
@@ -2135,23 +2056,160 @@ process_pkts_back(uint16_t port_back, uint16_t port_front,
+		fe->state = GK_DECLINED;
+		fe->u.declined.expire_at = now +
+			policy->params.declined.expire_sec * cycles_per_sec;
+		break;
+
+	case GK_BPF:
+		fe->state = GK_BPF;
+		fe->u.bpf.expire_at = now +
+			policy->params.bpf.expire_sec * cycles_per_sec;
+		fe->u.bpf.program_index = policy->params.bpf.program_index;
+		fe->u.bpf.cookie = policy->params.bpf.cookie;
+		break;
+
+	default:
+		GK_LOG(ERR, "Unknown flow state %u\n", policy->state);
+		break;
+	}
+}
+
+static void
+add_ggu_policy_bulk(struct ggu_policy **policies,
+	uint32_t *flow_hash_val_arr, int num_policies,
+	struct gk_instance *instance, struct gk_config *gk_conf)
+{
+	int i;
+	struct ip_flow *flow_arr[num_policies];
+	int done_lookups;
+	int32_t pos_arr[num_policies];
+
+	for (i = 0; i < num_policies; i++)
+		flow_arr[i] = &policies[i]->flow;
+
+	done_lookups = 0;
+	while (done_lookups < num_policies) {
+		int ret;
+		int r;
+		uint32_t num_keys = num_policies - done_lookups;
+		if (num_keys > RTE_HASH_LOOKUP_BULK_MAX)
+			num_keys = RTE_HASH_LOOKUP_BULK_MAX;
+
+		ret = rte_hash_lookup_bulk_with_hash(
+			instance->ip_flow_hash_table,
+			(const void **)&flow_arr[done_lookups],
+			(hash_sig_t *)&flow_hash_val_arr[done_lookups],
+			num_keys, &pos_arr[done_lookups]);
+		if (ret != 0) {
+			GK_LOG(NOTICE,
+				"failed to find multiple keys in the hash table at lcore %u\n",
+				rte_lcore_id());
+		}
+
+		r = done_lookups + num_keys;
+		for (i = done_lookups; i < r; i++) {
+			int pos = pos_arr[i];
+			struct flow_entry *fe;
+			struct ggu_policy *policy = policies[i];
+
+			if (pos >= 0)
+				fe = &instance->ip_flow_entry_table[pos];
+			else if (instance->num_scan_del > 0)
+				continue;
+			else {
+				struct gk_fib *fib = look_up_fib(
+					&gk_conf->lpm_tbl, &policy->flow);
+				if (fib == NULL || fib->action !=
+						GK_FWD_GRANTOR) {
+					/*
+					 * Drop this solicitation to add
+					 * a policy decision.
+					 */
+					char err_msg[128];
+					ret = snprintf(err_msg, sizeof(err_msg),
+						"gk: at %s initialize flow entry error",
+						__func__);
+					RTE_VERIFY(ret > 0 &&
+						ret < (int)sizeof(err_msg));
+					print_flow_err_msg(
+						&policy->flow, err_msg);
+					continue;
+				}
+
+				/*
+				 * The function add_ggu_policy_bulk() only
+				 * fills up GK_GRANTED and GK_DECLINED states.
+				 * So, it doesn't need to call
+				 * initialize_flow_entry().
+				 */
+				fe = add_new_flow_from_policy(fib,
+					policy, instance, gk_conf,
+					flow_hash_val_arr[i]);
+				if (fe == NULL)
+					continue;
+			}
+
+			update_flow_entry(fe, policy);
+		}
+
+		done_lookups += num_keys;
+	}
+}
+
 static void
 process_cmds_from_mailbox(
 	struct gk_instance *instance, struct gk_config *gk_conf)
 {
 	int i;
 	int num_cmd;
+	int num_policies = 0;
 	unsigned int mailbox_burst_size = gk_conf->mailbox_burst_size;
 	struct gk_cmd_entry *gk_cmds[mailbox_burst_size];
+	struct ggu_policy *policies[mailbox_burst_size];
+	uint32_t flow_hash_val_arr[mailbox_burst_size];
```

Comment:
Once `struct gk_add_policy` is available from the first patch, we can have a single array here.